data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconBlockHandler.java

```diff
@@ -13,71 +13,116 @@
-    Map<String, List<String>> queryParamMap = param.getQueryParamMap();
-    if (queryParamMap.containsKey(ROOT)) {
-      Bytes32 root = Bytes32.fromHexString(param.getQueryParam(ROOT));
-      return client.getStore() != null ? client.getStore().getBlock(root) : null;
-    }
+      final Map<String, List<String>> queryParamMap = ctx.queryParamMap();
+      if (ctx.queryParamMap().containsKey(ROOT)) {
+        final Bytes32 root = Bytes32.fromHexString(validateQueryParameter(queryParamMap, ROOT));
+        if (client.getStore() != null) {
+          final SignedBeaconBlock block = client.getStore().getSignedBlock(root);
+          if (block != null) {
+            ctx.result(jsonProvider.objectToJSON(new BeaconBlockResponse(block)));
+            return;
+          }
+          ctx.status(SC_NOT_FOUND);
+          return;
+        }
+      }
 
-    UnsignedLong slot;
-    if (queryParamMap.containsKey(EPOCH)) {
-      slot = compute_start_slot_at_epoch(UnsignedLong.valueOf(param.getQueryParam(EPOCH)));
-    } else if (queryParamMap.containsKey(SLOT)) {
-      slot = UnsignedLong.valueOf(param.getQueryParam(SLOT));
-    } else {
-      return null;
-    }
+      final UnsignedLong slot;
+      if (ctx.queryParamMap().containsKey(EPOCH)) {
+        slot =
+            compute_start_slot_at_epoch(
+                UnsignedLong.valueOf(validateQueryParameter(queryParamMap, EPOCH)));
+      } else if (ctx.queryParamMap().containsKey(SLOT)) {
+        slot = UnsignedLong.valueOf(validateQueryParameter(queryParamMap, SLOT));
+      } else {
+        throw new IllegalArgumentException(
+            "Query parameter missing. Must specify one of root or epoch or slot.");
+      }
 
-    return getBlockBySlot(slot)
-        .map(
-            block ->
-                ImmutableMap.of("block", block, "blockRoot", block.hash_tree_root().toHexString()))
-        .orElse(null);
+      final Optional<SignedBeaconBlock> blockBySlot = getBlockBySlot(slot);
+      if (blockBySlot.isPresent()) {
+        ctx.result(jsonProvider.objectToJSON(new BeaconBlockResponse(blockBySlot.get())));
+        return;
+      }
+      ctx.status(SC_NOT_FOUND);
+    } catch (final IllegalArgumentException e) {
+      ctx.result(jsonProvider.objectToJSON(new BadRequest(e.getMessage())));
+    }
   }
 
-  private Optional<BeaconBlock> getBlockBySlot(UnsignedLong slot) {
+  private Optional<SignedBeaconBlock> getBlockBySlot(final UnsignedLong slot) {
     return client
         .getBlockRootBySlot(slot)
-        .map(root -> client.getStore().getBlock(root))
-        .or(
-            () -> {
-              Optional<SignedBeaconBlock> signedBeaconBlock =
-                  historicalChainData.getFinalizedBlockAtSlot(slot).join();
-              return signedBeaconBlock.map(SignedBeaconBlock::getMessage);
-            });
-  };
+        .map(root -> client.getStore().getSignedBlock(root))
+        .or(() -> historicalChainData.getFinalizedBlockAtSlot(slot).join());
```

Comment:
Please avoid calling `join` in production code.  There are some places it can't be avoided and a few places that we do call it and shouldn't, but generally it's better to use the chaining methods (`thenApply`, `thenCompose` etc) and not just block the thread potentially forever.